In [1]:
# Import the relevant python libraries for the analysis
import pandas as pd
from pandas import DataFrame
import numpy as np
import statistics

In [2]:
# Load the dataset
mortalitad_materna = pd.read_csv('../data/mortalidad_materna.csv')
#materna.info()

### Create a **materna** subdataset *in English* from the **maternalidad_materna** dataset *in Spanish* with information including: 

- Residence Area
- Local Community Size
- Total Education Completed 
- Last Recorded Age
- Reason for Mortality
- Medical Assistance

### Step 1: Explore the Data and Create Clean Sub-Dataframe for Analysis

In [3]:
# 1. Combine patient birthdate information into one column
birth = DataFrame(mortalitad_materna, columns=['Año de nacimiento', 'Mes de nacimiento', 'Día de nacimiento'])
birth = mortalitad_materna['Año de nacimiento'].map(str) + '-' + mortalitad_materna['Mes de nacimiento'].map(str) + '-' + mortalitad_materna['Mes de nacimiento'].map(str)
print(birth.sort_values(ascending=True).head(2))
len(birth)

13088    0-0-0
12833    0-0-0
dtype: object


18163

In [4]:
# Combine patient date of death information into one column
death = DataFrame(mortalitad_materna, columns=['Año de la defunción', 'Mes de la defunción', 'Día de la defunción'])
death = mortalitad_materna['Año de la defunción'].map(str) + '-' + mortalitad_materna['Mes de la defunción'].map(str) + '-' + mortalitad_materna['Mes de la defunción'].map(str)
print(death.sort_values(ascending=True).head(2))
len(death)

17886       0-0-0
17516    1914-3-3
dtype: object


18163

#### Create variables for important location, economic, and educational factors associated with the instance of maternal mortality:
- Residence Area
- Local Community Size
- Total Education Completed 
- Last Recorded Age
- Reason for Mortality
- Medical Assistance

In [5]:
# Create variable to store:

#residence information 
residence_code = mortalitad_materna['Entidad de residencia']
residence_name = mortalitad_materna['Descripción de entidad de residencia']

#local community info
local_size = mortalitad_materna['Descripción del tamaño de localidad']
local_size_code = mortalitad_materna['Tamaño de localidad']

#educational level
edu_reached_code = mortalitad_materna['Escolaridad'] 
edu_reached = mortalitad_materna['Descripción de la escolaridad']

#age fulfilled by patient
last_age = mortalitad_materna['Edad cumplida']

#mortality reason
mortality_reason = mortalitad_materna['Razón de mortalidad materna']

#medical assistance
medical_received = mortalitad_materna['Descripción de la asistencia médica']

In [6]:
# Create a sub-dataframe to hold all date- information 
materna = pd.concat([birth, 
                   death, 
                   residence_code,
                   residence_name,
                   local_size,
                   local_size_code,
                   edu_reached_code,
                   edu_reached,
                   last_age,
                   mortality_reason,
                   medical_received], axis=1)
materna.columns = ['Date of Birth', 
                 'Date of Mortality', 
                 'Residence Code',
                 'Residence Name',
                 'Local Community Size',
                 'Local Size Code',
                 'Education Code',
                 'Education Completed',
                 'Age at Death',
                 'Reason for Mortality',
                 'Medical Assistance Received']
    
materna.head(2)

Date of Birth Date of Mortality  Residence Code Residence Name  \
0      1972-7-7          2002-3-3              29       Tlaxcala   
1      1967-4-4          2003-5-5              29       Tlaxcala   

         Local Community Size  Local Size Code  Education Code  \
0  40000 a 49999   habitantes               10               7   
1  40000 a 49999   habitantes               10               6   

                    Education Completed  Age at Death  Reason for Mortality  \
0                           PROFESIONAL            29                     1   
1  BACHILLERATO O PREPARATORIA COMPLETA            36                     1   

  Medical Assistance Received  
0         CON ATENCION MEDICA  
1         CON ATENCION MEDICA

### Data Wrangling: 

#### Part 1: Check for null or errors within materna

In [7]:
# Order dataframe to list in ascending order of approx. age at death
materna = materna.sort_values(by=['Age at Death'],ascending=True)
materna.head()

Date of Birth Date of Mortality  Residence Code  \
9598       1998-7-7          2009-7-7               7   
6407       1996-1-1        2008-12-12              30   
9084     2001-12-12          2014-5-5              27   
10883    1983-11-11          2002-6-6              15   
1266       1993-0-0          2005-1-1              12   

                        Residence Name           Local Community Size  \
9598                           Chiapas  1 a 999            habitantes   
6407   Veracruz de Ignacio de la Llave     50000 a 74999   habitantes   
9084                           Tabasco     250000 a 499999 habitantes   
10883                           México      2500 a 4999    habitantes   
1266                          Guerrero      2500 a 4999    habitantes   

       Local Size Code  Education Code    Education Completed  Age at Death  \
9598                 1               2    PRIMARIA INCOMPLETA            11   
6407                11               3      PRIMARIA COMPLETA            12   
9084                14               4  SECUNDARIA INCOMPLETA            12   
10883                4               5    SECUNDARIA COMPLETA            12   
1266                 4               2    PRIMARIA INCOMPLETA            12   

       Reason for Mortality Medical Assistance Received  
9598                      1         CON ATENCION MEDICA  
6407                      1         CON ATENCION MEDICA  
9084                      1         CON ATENCION MEDICA  
10883                     1         CON ATENCION MEDICA  
1266                      1         SIN ATENCION MEDICA

In [8]:
# Reset Index 
materna = materna.reset_index(drop=True)
materna.head()
materna.tail()

Date of Birth Date of Mortality  Residence Code       Residence Name  \
18158         0-0-0          1975-2-2              25              Sinaloa   
18159         0-0-0          2006-9-9              27              Tabasco   
18160         0-0-0          2000-6-6              16  Michoacán de Ocampo   
18161         0-0-0             0-0-0               0      No especificado   
18162         0-0-0          2002-4-4               8            Chihuahua   

                Local Community Size  Local Size Code  Education Code  \
18158  1 a 999            habitantes                1               0   
18159     75000 a 99999   habitantes               12               1   
18160               No Especificado.                0               0   
18161               No Especificado.                0               0   
18162  1 a 999            habitantes                1               2   

       Education Completed  Age at Death  Reason for Mortality  \
18158      NO ESPECIFICADO           998                     0   
18159              NINGUNA           998                     1   
18160      NO ESPECIFICADO           998                     0   
18161      NO ESPECIFICADO           998                     0   
18162  PRIMARIA INCOMPLETA           998                     1   

      Medical Assistance Received  
18158             NO ESPECIFICADO  
18159             NO ESPECIFICADO  
18160             NO ESPECIFICADO  
18161         SIN ATENCION MEDICA  
18162         CON ATENCION MEDICA

#### Check if all values make sense
- Case 1: It is biologically impossible to reach 998 years of age. These data entries appear to all have '0-0-0' in the 'Date of Birth' column and should therefore be removed from the sub-dataset, dates.
- Case 2: What do the '0' and '1' entires for 'Reason for Mortality' mean? If one of these keys is not associated with maternal death, then those entries should also be removed from the sub-dataset, dates.

In [9]:
# Remove rows with NaN / '0-0-0' values in Date of Birth
materna = materna[materna['Date of Birth'] != '0-0-0']
materna.tail()

Date of Birth Date of Mortality  Residence Code  \
18134    1933-12-12          2015-8-8               9   
18135    1933-11-11        2015-12-12              15   
18136      1933-8-8          2015-9-9              24   
18137      1930-1-1          2016-2-2              14   
18138      1930-8-8          2017-1-1              30   

                        Residence Name         Local Community Size  \
18134                 Distrito Federal  500000 a 999999  habitantes   
18135                           México   250000 a 499999 habitantes   
18136                  San Luis Potosí  500000 a 999999  habitantes   
18137                          Jalisco  500000 a 999999  habitantes   
18138  Veracruz de Ignacio de la Llave    1000 a 1999    habitantes   

       Local Size Code  Education Code  Education Completed  Age at Death  \
18134               15               3    PRIMARIA COMPLETA            81   
18135               14               2  PRIMARIA INCOMPLETA            82   
18136               15               2  PRIMARIA INCOMPLETA            82   
18137               15               2  PRIMARIA INCOMPLETA            86   
18138                2               2  PRIMARIA INCOMPLETA            86   

       Reason for Mortality Medical Assistance Received  
18134                     0         CON ATENCION MEDICA  
18135                     0         CON ATENCION MEDICA  
18136                     0         CON ATENCION MEDICA  
18137                     0         CON ATENCION MEDICA  
18138                     0         CON ATENCION MEDICA

In [10]:
materna.head()

Date of Birth Date of Mortality  Residence Code  \
0      1998-7-7          2009-7-7               7   
1      1996-1-1        2008-12-12              30   
2    2001-12-12          2014-5-5              27   
3    1983-11-11          2002-6-6              15   
4      1993-0-0          2005-1-1              12   

                    Residence Name           Local Community Size  \
0                          Chiapas  1 a 999            habitantes   
1  Veracruz de Ignacio de la Llave     50000 a 74999   habitantes   
2                          Tabasco     250000 a 499999 habitantes   
3                           México      2500 a 4999    habitantes   
4                         Guerrero      2500 a 4999    habitantes   

   Local Size Code  Education Code    Education Completed  Age at Death  \
0                1               2    PRIMARIA INCOMPLETA            11   
1               11               3      PRIMARIA COMPLETA            12   
2               14               4  SECUNDARIA INCOMPLETA            12   
3                4               5    SECUNDARIA COMPLETA            12   
4                4               2    PRIMARIA INCOMPLETA            12   

   Reason for Mortality Medical Assistance Received  
0                     1         CON ATENCION MEDICA  
1                     1         CON ATENCION MEDICA  
2                     1         CON ATENCION MEDICA  
3                     1         CON ATENCION MEDICA  
4                     1         SIN ATENCION MEDICA

**Case 1 Conclusion**: It appears that removing rows with '0-0-0' in the 'Date of Birth' column did remove all entries outside of the biological scope of maximum age at death. However, it is unlikely that a women who reached the age of 81 died due to maternal reasons since this age is outside the childe-bearing years. The analysis of the '0' and '1' in the 'Reason for Mortality' column may shed more light since all younger ages seem to be associated with '1' while older ages are associated with '0'. 

In [11]:
# Create a variable for the description of Reason for Mortality Description
mortality_description = mortalitad_materna['Descripción de la razón de mortalidad materna']

# Create a sub-dataframe to show interaction of Reason for Mortality Code and Description
mortality = pd.concat([mortality_reason, mortality_description], axis=1)
mortality.columns = ['Reason Mortality Code', 'Reason Mortality Description']
mortality.head()

Reason Mortality Code                       Reason Mortality Description
0                      1  Muertes Maternas para la razón de Mortalidad M...
1                      1  Muertes Maternas para la razón de Mortalidad M...
2                      1  Muertes Maternas para la razón de Mortalidad M...
3                      1  Muertes Maternas para la razón de Mortalidad M...
4                      1  Muertes Maternas para la razón de Mortalidad M...

In [12]:
mortality.tail()

Reason Mortality Code  \
18158                      0   
18159                      0   
18160                      0   
18161                      0   
18162                      0   

                            Reason Mortality Description  
18158  Muertes Maternas excluidas para la razón de Mo...  
18159  Muertes Maternas excluidas para la razón de Mo...  
18160  Muertes Maternas excluidas para la razón de Mo...  
18161  Muertes Maternas excluidas para la razón de Mo...  
18162  Muertes Maternas excluidas para la razón de Mo...

In [13]:
print('0 Description:')
print('Spanish: Muertes Maternas excluidas para la razón de Mortalidad Materna')
print('English: Maternal deaths excluded for the reason of Maternal Mortality')

0 Description:
Spanish: Muertes Maternas excluidas para la razón de Mortalidad Materna
English: Maternal deaths excluded for the reason of Maternal Mortality


In [14]:
print('1 Description:')
print('Spanish: Muertes Maternas para la razón de Mortalidad Materna')
print('English: Maternal deaths for the reason of Maternal Mortality')

1 Description:
Spanish: Muertes Maternas para la razón de Mortalidad Materna
English: Maternal deaths for the reason of Maternal Mortality


**Case 2 Conclusion**: Since '1' refers to recorded maternal-deaths and '0' refers to recorded deaths that are *not* associated with maternity, all rows containing '0' in the 'Reason for Mortality' column should be removed.

In [15]:
# Remove rows with 0 values in Reason for Mortality
materna = materna[materna['Reason for Mortality'] != 0 ]
materna.tail()

Date of Birth Date of Mortality  Residence Code Residence Name  \
18113      1960-3-3        2010-10-10              12       Guerrero   
18116      1960-1-1        2011-10-10              12       Guerrero   
18121    1963-12-12        2017-10-10              15         México   
18123      1961-8-8          2015-5-5               7        Chiapas   
18125    1942-12-12        2002-12-12              20         Oaxaca   

                Local Community Size  Local Size Code  Education Code  \
18113  1 a 999            habitantes                1               7   
18116  1 a 999            habitantes                1               0   
18121      10000 a 14999  habitantes                6               7   
18123      1000 a 1999    habitantes                2               1   
18125      2000 a 2499    habitantes                3               2   

       Education Completed  Age at Death  Reason for Mortality  \
18113          PROFESIONAL            50                     1   
18116      NO ESPECIFICADO            51                     1   
18121          PROFESIONAL            53                     1   
18123              NINGUNA            53                     1   
18125  PRIMARIA INCOMPLETA            54                     1   

      Medical Assistance Received  
18113         CON ATENCION MEDICA  
18116         CON ATENCION MEDICA  
18121         CON ATENCION MEDICA  
18123         CON ATENCION MEDICA  
18125         CON ATENCION MEDICA

#### Part 2: Translating Important Columns in **materna** from Spanish *using translation_english.txt*
- Important columns that need translation include: 
    - *Education Completed*
        - Translate Spanish Descriptions into Integer Values that Are Comparable
    - *Local Community Size*
        - Translate Spanish Descriptions into Integer Values that Are Comparable
    - *Medical Assistance Received*
        - Translate Medical Assistance Received into a Binary Column

**Education Completed**
- 9
    - 'POSGRADO' = Post-Graduate Education
- 8
    - 'PROFESIONAL' = Professional School
- 7
    - 'BACHILLERATO O PREPARATORIA COMPLETA' = High School (grades 10-12) complete
- 6
    - 'BACHILLERATO O PREPARATORIA INCOMPLETA' = High School (grades 10-12) incomplete
- 5 
    - 'PRIMARIA COMPLETA' = Elementary School (grades 1-6) complete 
- 4 
    - 'PRIMARIA INCOMPLETA' = Elementary School (grades 1-6) incomplete
- 3 
    - 'SECUNDARIA COMPLETA' = Junior High (grades 7-9) complete
- 2
    - 'SECUNDARIA INCOMPLETA' = Junior High (grades 7-9) incomplete
- 1
   - 'PREESCOLAR' = Preschool complete
-  0 
    - Combine the following entries:'SE IGNORA' = It was 'ignored' / 'NO ESPECIFICADO' = Not specified / 'NINGUNA' = NONE

In [16]:
# Create a sub-dataframe to show interaction of Education Code and Education Completed
education = materna[['Education Code', 'Education Completed']].sort_values(by='Education Code')
education = education.drop_duplicates()
print(len(education))
education

12


Education Code                     Education Completed
12865               0                         NO ESPECIFICADO
13582               1                                 NINGUNA
10135               2                     PRIMARIA INCOMPLETA
9046                3                       PRIMARIA COMPLETA
13586               4                   SECUNDARIA INCOMPLETA
2278                5                     SECUNDARIA COMPLETA
4727                6    BACHILLERATO O PREPARATORIA COMPLETA
11751               7                             PROFESIONAL
8506               10                                POSGRADO
10179              11  BACHILLERATO O PREPARATORIA INCOMPLETA
14959              12                              PREESCOLAR
4866               99                               SE IGNORA

In [17]:
# Overwriting column with replaced value of Education

# SE IGNORA / NINGUNA / NO ESPECIFICADO
materna["Education Completed"]= materna["Education Completed"].replace(['SE IGNORA', 'NINGUNA', 'NO ESPECIFICADO'], 0)

# PREESCOLAR
materna["Education Completed"]= materna["Education Completed"].replace('PREESCOLAR', 1)

# PRIMARIA
#INCOMPLETA
materna["Education Completed"]= materna["Education Completed"].replace('PRIMARIA INCOMPLETA', 2)
#COMPLETA
materna["Education Completed"]= materna["Education Completed"].replace('PRIMARIA COMPLETA', 3)

# SECUNDARIA
#INCOMPLETA
materna["Education Completed"]= materna["Education Completed"].replace('SECUNDARIA INCOMPLETA', 4)
#COMPLETA
materna["Education Completed"]= materna["Education Completed"].replace('SECUNDARIA COMPLETA', 5)

# BACHILLERATO O PREPARATORIA
#INCOMPLETA
materna["Education Completed"]= materna["Education Completed"].replace('BACHILLERATO O PREPARATORIA INCOMPLETA', 6)
#COMPLETA
materna["Education Completed"]= materna["Education Completed"].replace('BACHILLERATO O PREPARATORIA COMPLETA', 7)

# PROFESIONAL
materna["Education Completed"]= materna["Education Completed"].replace('PROFESIONAL', 8)

#POSGRADO
materna["Education Completed"]= materna["Education Completed"].replace('POSGRADO', 9)

In [18]:
# Test output
list(materna['Education Completed'].sort_values().unique())

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

**Local Community Size**

In [19]:
# Create a sub-dataframe to show interaction of Education Code and Education Completed
local_community = materna[['Local Size Code', 'Local Community Size']].sort_values(by='Local Size Code')
local_community = local_community.drop_duplicates()
print(len(local_community))
local_community

18


Local Size Code           Local Community Size
15499                0               No Especificado.
10827                1  1 a 999            habitantes
6759                 2      1000 a 1999    habitantes
16918                3      2000 a 2499    habitantes
4255                 4      2500 a 4999    habitantes
12797                5      5000 a 9999    habitantes
16719                6      10000 a 14999  habitantes
9995                 7     15000 a 19999   habitantes
10151                8     20000 a 29999   habitantes
5722                 9     30000 a 39999   habitantes
15396               10     40000 a 49999   habitantes
666                 11     50000 a 74999   habitantes
10491               12     75000 a 99999   habitantes
8608                13     100000 a 249999 habitantes
13247               14     250000 a 499999 habitantes
17432               15    500000 a 999999  habitantes
13697               16  1000000 a 1499999  habitantes
11270               17       1500000 y más habitantes

*Observations*: Local Community Size appears to have corresponding codes ordered from least to greatest community size already. Thereofore the 'Local Size Code' values can remain while the 'Local Community Size' column can be dropped.

#### Medical Assistance Received
- 0: WITH Medical Assistance
- 1: Unspecified/WITHOUT Medical Assistance

In [20]:
# Create a list item to hold comparison response
binary_medassist = []

# Create an iteration function to compare region mean to popupation mean
for medassist in materna['Medical Assistance Received']:
    
    #test for assistance
    if medassist == 'CON ATENCION MEDICA':
        binary_medassist.append(0)
    else:
        binary_medassist.append(1)

In [21]:
# Test output
#binary_medassist

In [22]:
# Convert the list to a Series and add as new column
materna['Received(0)/Not(1) Medical Assistance'] = pd.Series(binary_medassist)
materna.head()

Date of Birth Date of Mortality  Residence Code  \
0      1998-7-7          2009-7-7               7   
1      1996-1-1        2008-12-12              30   
2    2001-12-12          2014-5-5              27   
3    1983-11-11          2002-6-6              15   
4      1993-0-0          2005-1-1              12   

                    Residence Name           Local Community Size  \
0                          Chiapas  1 a 999            habitantes   
1  Veracruz de Ignacio de la Llave     50000 a 74999   habitantes   
2                          Tabasco     250000 a 499999 habitantes   
3                           México      2500 a 4999    habitantes   
4                         Guerrero      2500 a 4999    habitantes   

   Local Size Code  Education Code  Education Completed  Age at Death  \
0                1               2                    2            11   
1               11               3                    3            12   
2               14               4                    4            12   
3                4               5                    5            12   
4                4               2                    2            12   

   Reason for Mortality Medical Assistance Received  \
0                     1         CON ATENCION MEDICA   
1                     1         CON ATENCION MEDICA   
2                     1         CON ATENCION MEDICA   
3                     1         CON ATENCION MEDICA   
4                     1         SIN ATENCION MEDICA   

   Received(0)/Not(1) Medical Assistance  
0                                    0.0  
1                                    0.0  
2                                    0.0  
3                                    0.0  
4                                    1.0

In [23]:
# Drop columns that are unnecessary
materna = materna.drop(columns=['Date of Birth', 'Date of Mortality', 'Local Community Size','Medical Assistance Received', 'Education Code', 'Reason for Mortality'])
materna.head()

Residence Code                   Residence Name  Local Size Code  \
0               7                          Chiapas                1   
1              30  Veracruz de Ignacio de la Llave               11   
2              27                          Tabasco               14   
3              15                           México                4   
4              12                         Guerrero                4   

   Education Completed  Age at Death  Received(0)/Not(1) Medical Assistance  
0                    2            11                                    0.0  
1                    3            12                                    0.0  
2                    4            12                                    0.0  
3                    5            12                                    0.0  
4                    2            12                                    1.0

In [24]:
# Analyze shape of cleaned data
materna.describe()

Residence Code  Local Size Code  Education Completed  Age at Death  \
count    16636.000000     16636.000000         16636.000000  16636.000000   
mean        16.279574         7.810050             4.023684     28.356756   
std          7.850170         5.912597             2.387759      7.323171   
min          1.000000         0.000000             0.000000     11.000000   
25%         10.000000         2.000000             2.000000     22.000000   
50%         15.000000         6.000000             4.000000     28.000000   
75%         21.000000        14.000000             5.000000     34.000000   
max         34.000000        17.000000             9.000000     54.000000   

       Received(0)/Not(1) Medical Assistance  
count                           15302.000000  
mean                                0.142269  
std                                 0.349337  
min                                 0.000000  
25%                                 0.000000  
50%                                 0.000000  
75%                                 0.000000  
max                                 1.000000

In [25]:
print('There are '+ str(len(np.unique(materna['Residence Name']))) + ' Provinces in Mexico.')
list(np.unique(materna['Residence Name']))

There are 34 Provinces in Mexico.


['Aguascalientes',
 'Baja California',
 'Baja California Sur',
 'Campeche',
 'Chiapas',
 'Chihuahua',
 'Coahuila de Zaragoza',
 'Colima',
 'Distrito Federal',
 'Durango',
 'Estados Unidos de Norteamérica',
 'Guanajuato',
 'Guerrero',
 'Hidalgo',
 'Jalisco',
 'Michoacán de Ocampo',
 'Morelos',
 'México',
 'Nayarit',
 'Nuevo León',
 'Oaxaca',
 'Otros paises latinoamericanos',
 'Puebla',
 'Querétaro Arteaga',
 'Quintana Roo',
 'San Luis Potosí',
 'Sinaloa',
 'Sonora',
 'Tabasco',
 'Tamaulipas',
 'Tlaxcala',
 'Veracruz de Ignacio de la Llave',
 'Yucatán',
 'Zacatecas']

### Additional Cleaning of Data: 
There are 31 states and 1 federal entity in Mexico, so the length of the 'Regions' column should be 32, not 34. When translating the list of regions within the sub-dataset, it becomes apparent that not all entries are Provinces within Mexico, and need to be removed. Namely:
- Estados Unidos de Norteamérica - 'United States of America'
- Otros paises latinoamericanos - 'Other Latin American countries'
- No especificado - 'Not Specified'
- Otros paises - 'Other Countries'

In [26]:
# Remove unnecessary rows from region_ages sub-dataset
materna = materna[materna['Residence Name'] != 'Estados Unidos de Norteamérica' ]
materna = materna[materna['Residence Name'] != 'Otros paises latinoamericanos' ]
materna = materna[materna['Residence Name'] != 'No especificado' ]
materna = materna[materna['Residence Name'] != 'Otros paises' ]

print('There are '+ str(len(np.unique(materna['Residence Name']))) + ' Provinces in Mexico.')
list(np.unique(materna['Residence Name']))

There are 32 Provinces in Mexico.


['Aguascalientes',
 'Baja California',
 'Baja California Sur',
 'Campeche',
 'Chiapas',
 'Chihuahua',
 'Coahuila de Zaragoza',
 'Colima',
 'Distrito Federal',
 'Durango',
 'Guanajuato',
 'Guerrero',
 'Hidalgo',
 'Jalisco',
 'Michoacán de Ocampo',
 'Morelos',
 'México',
 'Nayarit',
 'Nuevo León',
 'Oaxaca',
 'Puebla',
 'Querétaro Arteaga',
 'Quintana Roo',
 'San Luis Potosí',
 'Sinaloa',
 'Sonora',
 'Tabasco',
 'Tamaulipas',
 'Tlaxcala',
 'Veracruz de Ignacio de la Llave',
 'Yucatán',
 'Zacatecas']

In [27]:
# Store as a global variable that can be uploaded to other Jupyter Notebooks
%store materna

Stored 'materna' (DataFrame)


### Preparation of Data for Machine Learning Analysis
- Create variables for age distribution by region
- Create a sub-dataframe for Machine Learning model

### Part 1: Create a sample region array variables to hold age distribution per region 

In [28]:
# Test code to create function
aqua = materna[materna['Residence Name'] == 'Aguascalientes']
aqua = aqua['Age at Death']
aqua = np.array(aqua)

mex = materna[materna['Residence Name'] == 'México']
mex = mex['Age at Death']
mex = np.array(mex)

print('Aguascalientes Sample Length: '+ str(aqua))
print('México Sample Length: ' + str(mex))

Aguascalientes Sample Length: [15 15 16 16 17 17 17 17 17 17 18 18 18 18 19 19 19 19 20 20 20 20 20 20
 20 21 21 21 21 22 22 22 22 22 22 23 23 23 23 23 23 24 24 24 25 25 25 25
 25 25 26 26 26 27 27 27 28 28 28 28 28 28 29 29 29 29 29 29 30 30 30 30
 30 30 31 31 31 31 31 31 31 32 32 32 32 33 33 33 33 33 33 34 34 34 35 35
 35 35 35 35 35 35 35 36 36 36 36 36 37 37 37 38 38 38 39 39 39 39 39 40
 40 41 42 42 42 43 43]
México Sample Length: [12 12 14 ... 49 49 53]


Since it appears that the sample size of ages of maternal death within the Provinces varies, the total sample per Province should be stored in unique age array variables. The process of creating the age array is repeatable, so a function should be created then applied to each Province. The array of ages variable can then be stored in a dictionary as a value with the associated Province as the key.

In [29]:
# Create a function to group all ages associated with materna death within a Province and store the ages in an array
def age_array(str):
    
    """Create arrays for all Ages of Maternal Death within a Region"""
    
    ages = materna[materna['Residence Name'] == str] # select the region 'str' from the 'Region' column
    ages = ages['Age at Death'] # select the ages within the region
    ages = np.array(ages) # store the ages in an array
    return ages # return the unique array

In [30]:
# Test output
print('Aguascalientes', age_array('Aguascalientes'))
print('México', age_array('México'))

Aguascalientes [15 15 16 16 17 17 17 17 17 17 18 18 18 18 19 19 19 19 20 20 20 20 20 20
 20 21 21 21 21 22 22 22 22 22 22 23 23 23 23 23 23 24 24 24 25 25 25 25
 25 25 26 26 26 27 27 27 28 28 28 28 28 28 29 29 29 29 29 29 30 30 30 30
 30 30 31 31 31 31 31 31 31 32 32 32 32 33 33 33 33 33 33 34 34 34 35 35
 35 35 35 35 35 35 35 36 36 36 36 36 37 37 37 38 38 38 39 39 39 39 39 40
 40 41 42 42 42 43 43]
México [12 12 14 ... 49 49 53]


In [31]:
# Create a variable for 'Region' names using np.unique()
list_regions = np.unique(materna['Residence Name'])

# Create an empty dictionary to hold the {Region : region_age_array} key pairs
age_by_state = {}

In [32]:
# Use the age_array function with iteration over residence to create the {Region : region_age_array} key pairs
for region in list_regions:
    age_by_state[region] = age_array(region) # add arrays as values in dictionary with region-key

In [33]:
# Test output
print('Aguascalientes', age_by_state['Aguascalientes'])

Aguascalientes [15 15 16 16 17 17 17 17 17 17 18 18 18 18 19 19 19 19 20 20 20 20 20 20
 20 21 21 21 21 22 22 22 22 22 22 23 23 23 23 23 23 24 24 24 25 25 25 25
 25 25 26 26 26 27 27 27 28 28 28 28 28 28 29 29 29 29 29 29 30 30 30 30
 30 30 31 31 31 31 31 31 31 32 32 32 32 33 33 33 33 33 33 34 34 34 35 35
 35 35 35 35 35 35 35 36 36 36 36 36 37 37 37 38 38 38 39 39 39 39 39 40
 40 41 42 42 42 43 43]


In [34]:
# Store as a global variable 
%store age_by_state

Stored 'age_by_state' (dict)


### Part 2: Create a sub-dataframe for the Machine Learning Model
- residence name
- residence code (index)
- region mean
- region sample size (n)
- region variance
- binary target
- mean educational level
- mean local community size
- mean presence of medical care

In [35]:
# Var for residence name 
residence_uniq = np.unique(materna['Residence Name'])

# Var for residence code
residence_code = np.unique(materna['Residence Code'])

# Create the sub-dateframe for region and region code
res_dataset = pd.DataFrame(residence_uniq, index=residence_code)
res_dataset = res_dataset.rename(columns={0:'Region'})
res_dataset.head()

Region
1       Aguascalientes
2      Baja California
3  Baja California Sur
4             Campeche
5              Chiapas

#### Calculate the Mean Age per Region

In [36]:
# Test Code
mean_death_list_trial = []

aguas = materna[materna['Residence Code'] == 1 ]
aguas = aguas[['Residence Code', 'Age at Death']]
aguas_mean = aguas['Age at Death'].mean()
aguas_mean = '{0:0.2f}'.format(aguas_mean)
print(aguas_mean)
mean_death_list_trial.append(aguas_mean)
print(mean_death_list_trial)

baja = materna[materna['Residence Code'] == 2 ]
baja = baja[['Residence Code', 'Age at Death']]
baja_mean = baja['Age at Death'].mean()
baja_mean = '{0:0.2f}'.format(baja_mean)
print(baja_mean)
mean_death_list_trial.append(baja_mean)
print(mean_death_list_trial)

28.36
['28.36']
27.15
['28.36', '27.15']


In [37]:
# Create an empty list to store region sample size and mean age of maternal death
region_mean = []
region_n = []

# Calculate the mean age of maternal death per region
for i in materna['Residence Code'].sort_values().unique():
    """Calculate Length of Age Array and Mean Age per Region"""

    sub_df = materna[materna['Residence Code'] == (i - 1)] # select one region
    n = len(sub_df['Age at Death']) # calculate sample length
    mean = sub_df['Age at Death'].mean() # calculate mean of region
    region_n.append(round(n, 2)) # append n to list
    region_mean.append(round(mean, 2)) # append mean to list

In [38]:
# Test output
print(region_mean[1], region_n[1], type(region_mean[1]))
print(region_mean[2], region_n[2], type(region_mean[2]))

28.36 127 <class 'numpy.float64'>
27.15 361 <class 'numpy.float64'>


In [39]:
# Convert the list to a Series and add as new column
res_dataset['μ Age Maternal Mortality'] = pd.Series(region_mean)
res_dataset['Region (n)'] = pd.Series(region_n)
res_dataset.tail()

Region  μ Age Maternal Mortality  Region (n)
28                       Tamaulipas                     28.26       377.0
29                         Tlaxcala                     28.45       178.0
30  Veracruz de Ignacio de la Llave                     27.94      1239.0
31                          Yucatán                     27.50       258.0
32                        Zacatecas                       NaN         NaN

*Preliminary Observation*: Since the region 'Zacatecas' has a NaN value for mean age of maternal death, the contents of 'Zacatecas' need to be adjusted from NaN to the actual mean of the data for the res_dataset. *With a quick reference, the initial values in the res_dataset match the individually calculated mean for both Aguascalientes and Baja California, so we know the NaN is not due to shifted values.*

#### Clean Data by Replacing NaN/Null values with the Correct Data

In [40]:
# Calculate the mean Age of Death for region 'Zacatecas'
zaca = materna[materna['Residence Code'] == 32 ]
zaca = zaca['Age at Death']

# Calculate sample size
zaca_n = len(zaca)

# Calculate mean
zaca_mean = zaca.mean()
zaca_mean = round(mean, 2)
print(zaca_mean)

# Change contents of res_dataset NaN to calculated mean
res_dataset['μ Age Maternal Mortality'] = res_dataset['μ Age Maternal Mortality'].replace(np.nan, zaca_mean)
res_dataset['Region (n)'] = res_dataset['Region (n)'].replace(np.nan, zaca_n)
res_dataset.tail()

27.5


Region  μ Age Maternal Mortality  Region (n)
28                       Tamaulipas                     28.26       377.0
29                         Tlaxcala                     28.45       178.0
30  Veracruz de Ignacio de la Llave                     27.94      1239.0
31                          Yucatán                     27.50       258.0
32                        Zacatecas                     27.50       191.0

#### Calculate the Age Variance by Region

In [41]:
# Test code
aguas = materna[materna['Residence Code'] == 1 ]
aguas = aguas[['Residence Code', 'Age at Death']]
aguas_var = statistics.pvariance(aguas['Age at Death'])
print('Aguas', aguas_var)

baja = materna[materna['Residence Code'] == 2 ]
baja = baja[['Residence Code', 'Age at Death']]
baja_var = statistics.pvariance(baja['Age at Death'])
print('Baja Cal', baja_var)

Aguas 55.16014632029264
Baja Cal 46.31196814020764


In [42]:
# Create an empty list to store age of maternal death variance per region
region_var = []

for i in materna['Residence Code'].sort_values().unique():
    """Calculate Age Standard Deviation and Age Variance per Region"""
    
    sub_df = materna[materna['Residence Code'] == i]
    age = list(sub_df['Age at Death'])
    var = statistics.pvariance(age) # calculate age variance of region pop
    
    for region in sub_df['Residence Name'].unique(): # prevent repeat entries in lists
        region_var.append(round(var, 2)) # append var to region list

In [43]:
# Test output - Make sure it matches Test Results
print('Test Results - Aguas', round(aguas_var,2))
print('Function Results - Aguas', region_var[0])

Test Results - Aguas 55.16
Function Results - Aguas 55.16


In [44]:
# Convert the list to a Series and add as new column
res_dataset['μ Age Variance'] = pd.Series(region_var, index=np.arange(1,33))
res_dataset.head()

Region  μ Age Maternal Mortality  Region (n)  μ Age Variance
1       Aguascalientes                     28.36       127.0           55.16
2      Baja California                     27.15       361.0           46.31
3  Baja California Sur                     27.56        66.0           55.37
4             Campeche                     26.87       126.0           44.29
5              Chiapas                     28.02       310.0           45.02

#### Create Binary Columns: 

μ Age 
- 0 : Region μ Age Maternal Mortality is *greater than or equal to* the population mean
- 1 : Region μ Age Maternal Mortality is *less than* the population mean

In [45]:
res_mean_age = res_dataset['μ Age Maternal Mortality'].mean()
print('The population mean is: ', round(res_mean_age, 2))

The population mean is:  28.16


In [46]:
# Create a dictionary item to hold comparison response
binary_mean = []

# Compare region mean to population mean
for mean in res_dataset['μ Age Maternal Mortality']:
    if mean >= res_mean_age:
        binary_mean.append(0)
    else:
        binary_mean.append(1)

In [47]:
# Test output
#binary_mean

In [48]:
# Convert the list to a Series and add as new column
res_dataset['Above(0) or Below(1) Average'] = pd.Series(binary_mean, index=np.arange(1,33))
res_dataset.head()

Region  μ Age Maternal Mortality  Region (n)  μ Age Variance  \
1       Aguascalientes                     28.36       127.0           55.16   
2      Baja California                     27.15       361.0           46.31   
3  Baja California Sur                     27.56        66.0           55.37   
4             Campeche                     26.87       126.0           44.29   
5              Chiapas                     28.02       310.0           45.02   

   Above(0) or Below(1) Average  
1                             0  
2                             1  
3                             1  
4                             1  
5                             1

#### Calculate Mean Educational Level per Region

In [49]:
# Create an empty list to store region sample size and mean age of maternal death
region_education = []
edu_dict = {}

# Create an iteration function to calculate the mean age of maternal death per region
for i in materna['Residence Code'].sort_values().unique():
    """Calculate Mean Education per Region"""

    sub_df = materna[materna['Residence Code'] == i ]
    region = str(sub_df['Residence Name'].unique())
    education = sub_df['Education Completed'].mean()
    mean_edu = round(education, 2)
    region_education.append(mean_edu)
    edu_dict[region] = mean_edu

In [50]:
# Test output
print(len(region_education))
#region_education
print(edu_dict)

# Store as a global variable
%store edu_dict

32
{"['Aguascalientes']": 4.56, "['Baja California']": 4.42, "['Baja California Sur']": 4.83, "['Campeche']": 4.12, "['Coahuila de Zaragoza']": 4.75, "['Colima']": 4.61, "['Chiapas']": 2.82, "['Chihuahua']": 3.34, "['Distrito Federal']": 5.28, "['Durango']": 3.96, "['Guanajuato']": 3.87, "['Guerrero']": 3.15, "['Hidalgo']": 4.18, "['Jalisco']": 4.32, "['México']": 4.45, "['Michoacán de Ocampo']": 3.64, "['Morelos']": 4.3, "['Nayarit']": 3.62, "['Nuevo León']": 4.8, "['Oaxaca']": 3.24, "['Puebla']": 3.76, "['Querétaro Arteaga']": 4.37, "['Quintana Roo']": 4.35, "['San Luis Potosí']": 4.04, "['Sinaloa']": 4.69, "['Sonora']": 4.96, "['Tabasco']": 4.6, "['Tamaulipas']": 4.47, "['Tlaxcala']": 4.82, "['Veracruz de Ignacio de la Llave']": 3.69, "['Yucatán']": 3.72, "['Zacatecas']": 4.35}
Stored 'edu_dict' (dict)


In [51]:
# Convert the list to a Series and add as new column
res_dataset['μ Region Education Level'] = pd.Series(region_education, index=np.arange(1,33))
res_dataset.tail()

Region  μ Age Maternal Mortality  Region (n)  \
28                       Tamaulipas                     28.26       377.0   
29                         Tlaxcala                     28.45       178.0   
30  Veracruz de Ignacio de la Llave                     27.94      1239.0   
31                          Yucatán                     27.50       258.0   
32                        Zacatecas                     27.50       191.0   

    μ Age Variance  Above(0) or Below(1) Average  μ Region Education Level  
28           45.45                             0                      4.47  
29           55.16                             0                      4.82  
30           53.60                             1                      3.69  
31           54.89                             1                      3.72  
32           48.10                             1                      4.35

#### Calculate Mean Educational Level per Region

In [52]:
# Create an empty list to store region sample size and mean age of maternal death
region_community_size = []
size_dict = {}

# Create an iteration function to calculate the mean age of maternal death per region
for i in materna['Residence Code'].sort_values().unique():
    """Calculate Mean Local Community Size per Region"""

    sub_df = materna[materna['Residence Code'] == i ]
    region = str(sub_df['Residence Name'].unique())
    local_community = sub_df['Local Size Code'].mean()
    mean_size = round(local_community, 2)
    region_community_size.append(mean_size)
    size_dict[region] = mean_size

In [53]:
# Test output
print(len(region_community_size))
#region_education
print(size_dict)

# Store as a global variable
%store size_dict

32
{"['Aguascalientes']": 10.42, "['Baja California']": 13.35, "['Baja California Sur']": 8.5, "['Campeche']": 6.56, "['Coahuila de Zaragoza']": 11.22, "['Colima']": 9.94, "['Chiapas']": 4.06, "['Chihuahua']": 7.83, "['Distrito Federal']": 12.5, "['Durango']": 6.97, "['Guanajuato']": 7.95, "['Guerrero']": 4.74, "['Hidalgo']": 4.61, "['Jalisco']": 10.82, "['México']": 10.18, "['Michoacán de Ocampo']": 7.12, "['Morelos']": 7.37, "['Nayarit']": 4.3, "['Nuevo León']": 13.2, "['Oaxaca']": 3.18, "['Puebla']": 6.77, "['Querétaro Arteaga']": 7.68, "['Quintana Roo']": 10.53, "['San Luis Potosí']": 6.12, "['Sinaloa']": 8.51, "['Sonora']": 10.23, "['Tabasco']": 5.36, "['Tamaulipas']": 11.52, "['Tlaxcala']": 5.9, "['Veracruz de Ignacio de la Llave']": 5.1, "['Yucatán']": 7.71, "['Zacatecas']": 5.84}
Stored 'size_dict' (dict)


In [54]:
# Convert the list to a Series and add as new column
res_dataset['μ Region Local Community Size'] = pd.Series(region_community_size, index=np.arange(1,33))
res_dataset.tail()

Region  μ Age Maternal Mortality  Region (n)  \
28                       Tamaulipas                     28.26       377.0   
29                         Tlaxcala                     28.45       178.0   
30  Veracruz de Ignacio de la Llave                     27.94      1239.0   
31                          Yucatán                     27.50       258.0   
32                        Zacatecas                     27.50       191.0   

    μ Age Variance  Above(0) or Below(1) Average  μ Region Education Level  \
28           45.45                             0                      4.47   
29           55.16                             0                      4.82   
30           53.60                             1                      3.69   
31           54.89                             1                      3.72   
32           48.10                             1                      4.35   

    μ Region Local Community Size  
28                          11.52  
29                           5.90  
30                           5.10  
31                           7.71  
32                           5.84

#### Calculate Mean Presence of Medical Care per Region

In [55]:
# Create an empty list to store region sample size and mean age of maternal death
region_medical = []
medical_dict = {}

# Create an iteration function to calculate the mean age of maternal death per region
for i in materna['Residence Code'].sort_values().unique():
    """Calculate Mean Education per Region"""
    
    sub_df = materna[materna['Residence Code'] == i ]
    region = str(sub_df['Residence Name'].unique())
    med_assist = round(sub_df['Received(0)/Not(1) Medical Assistance'].mean(), 2)
    region_medical.append(med_assist) 
    medical_dict[region] = med_assist

In [56]:
# Test output
print(len(region_medical))
#region_medical

# Store as a global variable
%store medical_dict

32
Stored 'medical_dict' (dict)


In [57]:
# Convert the list to a Series and add as new column
res_dataset['μ Presence(0)/Not(1) of Med Assist'] = pd.Series(region_medical, index=np.arange(1,33))
res_dataset.head()

Region  μ Age Maternal Mortality  Region (n)  μ Age Variance  \
1       Aguascalientes                     28.36       127.0           55.16   
2      Baja California                     27.15       361.0           46.31   
3  Baja California Sur                     27.56        66.0           55.37   
4             Campeche                     26.87       126.0           44.29   
5              Chiapas                     28.02       310.0           45.02   

   Above(0) or Below(1) Average  μ Region Education Level  \
1                             0                      4.56   
2                             1                      4.42   
3                             1                      4.83   
4                             1                      4.12   
5                             1                      4.75   

   μ Region Local Community Size  μ Presence(0)/Not(1) of Med Assist  
1                          10.42                                0.16  
2                          13.35                                0.16  
3                           8.50                                0.16  
4                           6.56                                0.14  
5                          11.22                                0.12

In [58]:
# Store as a global variable 
%store res_dataset

Stored 'res_dataset' (DataFrame)
